theano/tensor/nnet/nnet.py

```diff
@@ -600,34 +695,37 @@ class LogSoftmax(gof.Op):
     LogSoftmax activation function
     :math:`\\varphi(\\mathbf{x})_j =
     \\e^{(\mathbf{x}_j - log{\sum_{k=1}^K e^{\mathbf{x}_k})}}
-    where :math:`K` is the total number of neurons in the layer. This
-    activation function gets applied row-wise.
-
+    where :math:`K` is the total number of neurons in the layer.  This
+    activation function gets applied on the selected axis (Default axis is the last).
+    This function get applied when axis is a scalar and uses a fast C optimization.
     """
+    nin = 1
+    nout = 1
     __props__ = ()
+    params_type = ParamsType(axis=Scalar('int8'))
+
+    def __init__(self, axis=-1):
+        self.axis = axis
+        self.softmax_op = Softmax(self.axis)
 
     def make_node(self, x):
         x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
-        if x.ndim == 1:
-            x = tensor.shape_padleft(x, n_ones=1)
-
+        if x.type.dtype not in tensor.float_dtypes:
+            raise ValueError('x must be tensor of floats. Got %s' % x.type)
         return Apply(self, [x], [x.type()])
 
-    def perform(self, node, input_storage, output_storage):
+    def perform(self, node, input_storage, output_storage, param):
         x, = input_storage
-        xdev = x - x.max(axis=1)[:, None]
-        lsm = xdev - np.log(np.sum(np.exp(xdev), axis=1,
-                            keepdims=True))
+        # Make axis tuple or scalar
+        axes = param.axis
+        xdev = x - x.max(axis=axes, keepdims=True)
+        lsm = xdev - np.log(np.sum(np.exp(xdev), axis=axes, keepdims=True))
         output_storage[0][0] = lsm
 
     def grad(self, inp, grads):
         x, = inp
-        sm = softmax_op(x)
-        return [grads[0] - tensor.sum(grads[0], axis=1, keepdims=True) * sm]
+        sm = self.softmax_op(x)
+        return [grads[0] - tensor.sum(grads[0], axis=self.axis, keepdims=True) * sm]
```

Comment:
@ReyhaneAskari said I shouldn't do that because for each call at this function, we'll create a new object. So this I why, I was calling the Softmax into the init.